src/core/meshcop/border_agent.cpp

```diff
@@ -317,6 +337,217 @@ template <> void BorderAgent::HandleTmf<kUriRelayRx>(Coap::Message &aMessage, co
     FreeMessageOnError(message, error);
 }
 
+void BorderAgent::NotifyMeshCoPServiceChange(void)
+{
+    // We know all the TXT entries from OT core and the total size is less than 128.
+    static constexpr uint16_t kMaxTxtDataLen = 128;
+
+    uint8_t  txtDataBuffer[kMaxTxtDataLen];
+    uint16_t txtDataLen = kMaxTxtDataLen;
+
+    VerifyOrExit(mMeshCoPServiceChangedCallback.IsSet());
+    SuccessOrExit(mMeshCopTxtEncoder.EncodeTxtData(txtDataBuffer, txtDataLen));
+
+    mMeshCoPServiceChangedCallback.Invoke(txtDataBuffer, txtDataLen);
+
+exit:
+    return;
+}
+
+void BorderAgent::PostNotifyMeshCoPServiceChangeTask(void)
+{
+    if (mMeshCoPServiceChangedCallback.IsSet())
+    {
+        mNotifyMeshCoPServiceChangeTask.Post();
+    }
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+// BorderAgent::MeshCopTxtEncoder
+
+Error BorderAgent::MeshCopTxtEncoder::AppendTxtEntryTo(Appender   &aAppender,
+                                                       const char *aKey,
+                                                       const void *aValue,
+                                                       uint16_t    aValueLength)
+{
+    Dns::TxtEntry txtEntry;
+
+    txtEntry.Init(aKey, reinterpret_cast<const uint8_t *>(aValue), aValueLength);
+    return txtEntry.AppendTo(aAppender);
+}
+
+template <>
+Error BorderAgent::MeshCopTxtEncoder::AppendTxtEntryTo<NameData>(Appender       &aAppender,
+                                                                 const char     *aKey,
+                                                                 const NameData &aObject)
+{
+    return AppendTxtEntryTo(aAppender, aKey, aObject.GetBuffer(), aObject.GetLength());
+}
+
+Error BorderAgent::MeshCopTxtEncoder::EncodeTxtData(uint8_t *aOutput, uint16_t &aLength)
+{
+#if OPENTHREAD_CONFIG_THREAD_VERSION == OT_THREAD_VERSION_1_1
+    static const char kThreadVersionString[] = "1.1.1";
+#elif OPENTHREAD_CONFIG_THREAD_VERSION == OT_THREAD_VERSION_1_2
+    static const char kThreadVersionString[] = "1.2.0";
+#elif OPENTHREAD_CONFIG_THREAD_VERSION == OT_THREAD_VERSION_1_3
+    static const char kThreadVersionString[] = "1.3.0";
+#elif OPENTHREAD_CONFIG_THREAD_VERSION == OT_THREAD_VERSION_1_3_1
+    static const char kThreadVersionString[] = "1.3.1";
+#elif OPENTHREAD_CONFIG_THREAD_VERSION == OT_THREAD_VERSION_1_4
+    static const char kThreadVersionString[] = "1.4.0";
+#endif
```

Comment:
It is fine here for now, but we can later move this to a common place (maybe defining the `kThreadVersionAsStringp[]` in `version.hpp` and then defining it in `common/version.cpp`. Can do it later in its own small PR.

Reply:
SG! I'll do it.